src/json-observe.ts

```diff
@@ -0,0 +1,262 @@
+          }
+
+          if (typeof receiver === 'undefined') {
+            if (target.hasOwnProperty(key))  //exists
+            {
+              // when array element is set to `undefined`, should generate replace to `null` 
+              if (_isArray(target)) {
+                //undefined array elements are JSON.stringified to `null`
+                instance.defaultCallback({ op: 'replace', path: distPath, value: null });
+                return Reflect.set(target, key, receiver);
+              }
+              else {
+                instance.defaultCallback({ op: 'remove', path: distPath });
+                return Reflect.set(target, key, receiver);
+              }
+            }
+            //when new property is added, and set to `undefined`, nothing should be generated (undefined is JSON.stringified to no value
+            else if (!_isArray(target)) {
+              return Reflect.set(target, key, receiver);
+            }
+            //else use the origian settor below
+          }
+
+          if (_isArray(target) && !isInteger(key.toString())) //array attribute, don't intercept it
+          {
+            return Reflect.set(target, key, receiver);
+          }
+
+          if (target.hasOwnProperty(key))  //exists
+          {
+            if (typeof target[key] === 'undefined') {
+              if (_isArray(target)) {
+                instance.defaultCallback({ op: 'replace', path: distPath, value: receiver });
+                return Reflect.set(target, key, receiver);
+              }
+              //when `undefined` property is set to something, should generate add (undefined in JSON.stringified to no value)
+              else //object
+              {
+                instance.defaultCallback({ op: 'add', path: distPath, value: receiver });
+                return Reflect.set(target, key, receiver);
+              }
+            }
+            else {
+              instance.defaultCallback({ op: 'replace', path: distPath, value: receiver });
+              return Reflect.set(target, key, receiver);
+            }
+          }
+          else
+            instance.defaultCallback({ op: 'add', path: distPath, value: receiver });
+
+          return Reflect.set(target, key, receiver);
+        },
+        deleteProperty: function (target, key) {
+
+          //when when an `undefined` property is deleted
+          if (typeof target[key] === 'undefined')
+            return Reflect.deleteProperty(target, key);
+
+          // else {
+          var distPath = path + '/' + instance.escapePathComponent(key.toString());
+          instance.defaultCallback({ op: 'remove', path: distPath });
+          return Reflect.deleteProperty(target, key);
+        }
+      });
+    return proxy;
+  }
+
+  //grab tree's leaves one by one, encapsulate them into a proxy and return
+  private _proxifyObjectTreeRecursively(root: any, path): any {
+    if(!path) path = "";
```

Comment:
Do we need that line? we call `_proxifyObjectTreeRecursively` with second parameter in all the cases, and it's yet another `if` checked `n` times.